### DISCUSSION TITLE
How to Create a Many-to-Many relation but with one identifier

### DISCUSSION THREAD

**Discussion by caiovsoares** [prisma/prisma#18456] [Q&A] [answered]
I need to connect a room to another room, so that it's a two-way connection, so I can query for all the connections of a room.
To solve my problem, on create a connection, both the rooms need to know that connection.

Here is an example of something that would solve my problem. But it's wrong

```Prisma
model Room {
  id   String @id @default(uuid())
  name String

  RoomConnection  RoomConnection[] @relation("connection")
}

model RoomConnection {
  roomId  String
  roomId2 String
  room    Room   @relation("connection", fields: [roomId, roomId2], references: [id, id])

  @@id(fields: [roomId, roomId2])
}
```
So a room would have a variety of connections and each connection would be linked to two room ids.

**Comment by ludralph** [comment, marked as answer]
Hi @caiovsoares 👋 

I've developed this model based on what you are trying to achieve. 
```js
model Room {
  id                 String               @id @default(uuid())
  name               String
  connections        RoomConnection[]     @relation("RoomConnectionLinkFrom")
  connectedTo        RoomConnection[]     @relation("RoomConnectionLinkTo")
  RoomConnectionLink RoomConnectionLink[]
}

model RoomConnection {
  id                 String               @id @default(uuid())
  connection         String
  rooms              Room[]               @relation("RoomConnectionLinkFrom")
  connectedTo        Room[]               @relation("RoomConnectionLinkTo")
  RoomConnectionLink RoomConnectionLink[]
}

model RoomConnectionLink {
  id           String         @id @default(uuid())
  room         Room           @relation(fields: [roomId], references: [id])
  roomId       String
  connection   RoomConnection @relation(fields: [connectionId], references: [id])
  connectionId String

  @@unique([roomId, connectionId])
}
```
A `Room` has a many-to-many relationship with `RoomConnection` through the `RoomConnectionLink` model. `RoomConnectionLink` contains foreign keys to both `Room` and `RoomConnection`, and it represents a single connection between two rooms.

`Room` has two relations to `RoomConnection` - `connections` and `connectedTo`. `connections` represents the connections that start from this room, while `connectedTo` represents the connections that end in this room.
`RoomConnection` has a relation to `Room` called `rooms`, which represents the rooms that are connected by this connection.
`RoomConnectionLink specifies the relationships between `Room` and `RoomConnection` using the fields and references arguments of the @relation directive. The `roomId` and `connectionId` fields in RoomConnectionLink specify the foreign keys to Room and RoomConnection, respectively. The @@unique directive ensures that each connection between two rooms is unique.

To create a two-way connection between two rooms, you would need to create a `RoomConnection` object that references both of the rooms that are being connected.

Here's an example of how you might create a two-way connection between two rooms using this data model:
```js
// Create the two rooms that will be connected
const room1 = await prisma.room.create({
  data: {
    name: "Room 1",
  },
});

const room2 = await prisma.room.create({
  data: {
    name: "Room 2",
  },
});

// Create a new RoomConnection object that connects the two rooms
const connection = await prisma.roomConnection.create({
  data: {
    connection: "Connection 1",
    rooms: {
      connect: [
        { id: room1.id },
        { id: room2.id },
      ],
    },
  },
});

// Add the new RoomConnection object to each of the rooms' connections and connectedTo fields
await prisma.room.update({
  where: { id: room1.id },
  data: {
    connections: { connect: { id: connection.id } },
    connectedTo: { connect: { id: connection.id } },
  },
});

await prisma.room.update({
  where: { id: room2.id },
  data: {
    connections: { connect: { id: connection.id } },
    connectedTo: { connect: { id: connection.id } },
  },
});
```
To view all connected rooms, you can carry out this query
```js
const room = await prisma.room.findUnique({
    where: { id: roomId },
    include: {
      connections: {
        include: {
          connectedTo: true
        },
      },
    },
  });
 console.dir( room, { depth: Infinity } )
```
You would get an out like below
```js
{
  id: '204cf1e5-2dc1-4b16-9d14-23e59522d2fc',
  name: 'Room 1',
  connections: [
    {
      id: '25787ba8-53ec-4b9d-aece-9c678c367819',
      connection: 'Connection 1',
      connectedTo: [
        { id: '204cf1e5-2dc1-4b16-9d14-23e59522d2fc', name: 'Room 1' },
        { id: '10761e67-056f-4196-9b9e-0263bee153b2', name: 'Room 2' }
      ]
    }
  ]
}
```
Let me know if this helps

**Reply by ludralph** [reply]
Hi @caiovsoares 

Do you have any more questions around this issue? If no, it would be really helpful if you could quickly **mark this Discussion as _answered_** to indicate that your issue has been resolved. 

In case you do have any outstanding questions, feel free to let us know and we'd be happy to help!

**Reply by caiovsoares** [reply]
I think that is the best the solution evailable. It's not quite what I need, but solves the problem. It looks like Prisma does not supports that connection type yet.
Thank you too much for the help @ludralph !

**Comment by nikolasburk** [comment]
Hey @caiovsoares 👋 out of curiosity, did you consider an [implicit self-relationship](https://www.prisma.io/docs/concepts/components/prisma-schema/relations/self-relations#many-to-many-self-relations) for your data model?

It would look like this in your case I believe:

```prisma
model Room {
  id   String @id @default(uuid())
  name String

  connectedToRooms   Room[] @relation("RoomConnection")
  connectedFromRooms Room[] @relation("RoomConnection")
}
```

The benefit would be that your Prisma Client queries will be simpler because you save the extra hop through the `RoomConnection` model.

Also, you mentioned about your data model:

> Here is an example of something that would solve my problem. But it's wrong

Can you elaborate what exactly is wrong there?
